README.rst
This is the recaptcha widget for `yafowil <http://pypi.python.org/pypi/yafowil>`_,
based on `reCAPTCHA <http://www.google.com/recaptcha/intro/index.html>`_

- `Documentation <http://docs.yafowil.info/blueprints.html#recaptcha>`_

- `DEMO - see it Live <http://demo.yafowil.info/++widget++yafowil.widget.recaptcha/index.html>`_


Source Code
===========

The sources are in a GIT DVCS with its main branches at
`github <http://github.com/bluedynamics/yafowil.widget.recaptcha>`_.

We'd be happy to see many forks and pull-requests to make YAFOWIL even better.


Contributors
============

- Robert Niederreiter

- Georg Bernhard


CHANGES.rst
Changes
=======

1.3 (unreleased)
----------------

- No changes yet.


1.2 (2018-07-16)
----------------

- Python 3 compatibility.
  [rnix]

- Convert doctests to unittests.
  [rnix]


1.1 (2017-03-01)
----------------

- Use ``yafowil.utils.entry_point`` decorator.
  [rnix, 2016-06-28]


1.0 (2015-01-23)
----------------

- Make it work.
  [rnix]


src/yafowil/widget/recaptcha/tests.py
from node.tests import patch
from yafowil.base import ExtractionError
from yafowil.base import factory
from yafowil.compat import IS_PY2
from yafowil.tests import fxml
from yafowil.tests import YafowilTestCase
from yafowil.widget.recaptcha import widget as recaptcha_widget
import unittest
import yafowil.loader  # noqa


if not IS_PY2:
    from importlib import reload


class DummyRequest(dict):
    @property
    def request(self):
        return self


class RecaptchaResult(object):
    is_valid = False
    error_code = 'incorrect-captcha-sol'


recaptcha_result = RecaptchaResult()


def mock_submit(challenge_field, response_field, private_key, remote_addr):
    return recaptcha_result


class TestRecaptchaWidget(YafowilTestCase):

    def setUp(self):
        super(TestRecaptchaWidget, self).setUp()
        from yafowil.widget.recaptcha import widget
        reload(widget)

    def test_edit_renderer(self):
        # Render map widget with defaults
        widget = factory(
            'recaptcha',
            name='recaptcha',
            props={
                'public_key': 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
                'private_key': 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
                'lang': 'de',
                'theme': 'clean',
            })
        self.check_output("""
        <div>
          <div class="recaptcha" data-lang="de"
               data-public_key="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
               data-theme="clean" id="recaptcha-recaptcha">
          </div>
          <noscript>
            <iframe src="http://www.google.com/recaptcha/api/noscript?k=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
                    height="300" width="500" frameborder="0">
            </iframe>
            <br/>
            <textarea name="recaptcha_challenge_field" rows="3" cols="40"/>
            <input type="hidden" name="recaptcha_response_field"
                   value="manual_challenge"/>
          </noscript>
        </div>
        """, fxml('<div>\n' + widget() + '</div>'))

    def test_display_renderer(self):
        # Display renderer is not implemented
        widget = factory('recaptcha', 'recaptcha', mode='display')
        self.assertEqual(widget(), None)

    @patch(recaptcha_widget, 'submit', mock_submit)
    def test_extraction(self):
        # Widget extraction
        widget = factory(
            'recaptcha',
            name='recaptcha',
            props={
                'public_key': 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
                'private_key': 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
                'lang': 'de',
                'theme': 'clean',
            })

        request = DummyRequest()
        data = widget.extract(request)
        self.assertEqual(
            data.errors,
            [ExtractionError('The CAPTCHA solution was incorrect.')]
        )

        recaptcha_result.is_valid = True
        request = DummyRequest()
        request['recaptcha_challenge_field'] = '1234'
        request['recaptcha_response_field'] = 'manual_challenge'
        data = widget.extract(request)
        self.assertEqual(data.errors, [])


if __name__ == '__main__':
    unittest.main()


i18n.sh
#!/bin/bash
# Author: Robert Niederreiter <rnix@squarewave.at>
# License: no license
# Date: 2014-07-20
#
# Requirements:
#     - python lingua (sudo pip install lingua)
#     - gettext
#
# Usage:
#     Initial catalog creation (lang is the language identifier):
#         ./i18n.sh lang
#     Updating translation and compile catalog:
#         ./i18n.sh

# configuration
DOMAIN="yafowil.widget.recaptcha"
SEACH_PATH=src/yafowil/widget/recaptcha
LOCALES_PATH=src/yafowil/widget/recaptcha/locales
# end configuration

# create locales folder if not exists
if [ ! -d "$LOCALES_PATH" ]; then
    echo "Locales directory not exists, create"
    mkdir -p $LOCALES_PATH
fi

# create pot if not exists
if [ ! -f $LOCALES_PATH/$DOMAIN.pot ]; then
	echo "Create pot file"
	touch $LOCALES_PATH/$DOMAIN.pot
fi

# no arguments, extract and update
if [ $# -eq 0 ]; then
    echo "Extract messages"
    pot-create $SEACH_PATH -o $LOCALES_PATH/$DOMAIN.pot

    echo "Update translations"
    for po in $LOCALES_PATH/*/LC_MESSAGES/$DOMAIN.po; do
        msgmerge -o $po $po $LOCALES_PATH/$DOMAIN.pot
    done

    echo "Compile message catalogs"
    for po in $LOCALES_PATH/*/LC_MESSAGES/*.po; do
        msgfmt -o ${po%.*}.mo $po
    done

# first argument represents language identifier, create catalog
else
    cd $LOCALES_PATH
    mkdir -p $1/LC_MESSAGES
    msginit -i $DOMAIN.pot -o $1/LC_MESSAGES/$DOMAIN.po -l $1
fi


src/yafowil/widget/recaptcha/example.py
from yafowil.base import factory
import os


DOC_RECAPTCHA = """
reCAPTCHA
---------

reCAPTCHA widget.

.. code-block:: python

    recaptcha_public_key = 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
    recaptcha_private_key = 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
    captcha = factory('#field:recaptcha', props={
        'label': 'Verify',
        'public_key': recaptcha_public_key,
        'private_key': recaptcha_private_key,
        'lang': 'de',
        'theme': 'clean',
        'error.position': 'after',
    })
"""


def recaptcha_example():
    form = factory('fieldset', name='yafowil.widget.recaptcha.recaptcha')
    recaptcha_public_key = os.environ['recaptcha_public_key']
    recaptcha_private_key = os.environ['recaptcha_private_key']
    form['captcha'] = factory('#field:recaptcha', props={
        'label': 'Verify',
        'public_key': recaptcha_public_key,
        'private_key': recaptcha_private_key,
        'lang': 'de',
        'theme': 'clean',
        'error.position': 'after',
    })
    return {
        'widget': form,
        'doc': DOC_RECAPTCHA,
        'title': 'reCAPTCHA',
    }


def get_example():
    return [
        recaptcha_example(),
    ]


src/yafowil/widget/recaptcha/resources/widget.js
/*
 * yafowil recaptcha widget
 *
 * Requires: reCAPTCHA AJAX API
 * Optional: bdajax
 */

if (typeof(window.yafowil) == "undefined") yafowil = {};

(function($) {

    $(document).ready(function() {
        // initial binding
        yafowil.recaptcha.binder();

        // add after ajax binding if bdajax present
        if (typeof(window.bdajax) != "undefined") {
            $.extend(bdajax.binders, {
                recaptcha_binder: yafowil.recaptcha.binder
            });
        }
    });

    $.extend(yafowil, {

        recaptcha: {

            binder: function(context) {
                $('div.recaptcha', context).each(function() {
                    var elem = $(this);
                    var elem_id = elem.attr('id');
                    var public_key = elem.data('public_key');
                    var theme = elem.data('theme');
                    var lang = elem.data('lang');
                    Recaptcha.create(
                        public_key,
                        elem_id,
                        {
                            theme: theme,
                            lang: lang
                        }
                    );
                });
            },
        }
    });

})(jQuery);


src/yafowil/widget/recaptcha/widget.py
from recaptcha.client.captcha import submit
from yafowil.base import ExtractionError
from yafowil.base import factory
from yafowil.common import generic_extractor
from yafowil.tsf import TSF
from yafowil.utils import attr_value
from yafowil.utils import cssclasses
from yafowil.utils import cssid
from yafowil.utils import data_attrs_helper
from yafowil.utils import managedprops


_ = TSF('yafowil.widget.recaptcha')


error_messages = dict()
error_messages['invalid-site-private-key'] = _(
    'invalid-site-private-key',
    default='We weren\'t able to verify the private key.')
error_messages['invalid-request-cookie'] = _(
    'invalid-request-cookie',
    default='The challenge parameter of the verify script was incorrect.')
error_messages['incorrect-captcha-sol'] = _(
    'incorrect-captcha-sol',
    default='The CAPTCHA solution was incorrect.')
error_messages['captcha-timeout'] = _(
    'captcha-timeout',
    default='The solution was received after the CAPTCHA timed out.')
error_messages['recaptcha-not-reachable'] = _(
    'recaptcha-not-reachable',
    default='Unable to contact the reCAPTCHA verify server.')


NO_SCRIPT_TEMPLATE = """
<noscript>
   <iframe src="http://www.google.com/recaptcha/api/noscript?k={public_key}"
           height="300"
           width="500"
           frameborder="0">
   </iframe>
   <br />
   <textarea name="recaptcha_challenge_field"
             rows="3"
             cols="40"></textarea>
   <input type="hidden"
          name="recaptcha_response_field"
          value="manual_challenge" />
</noscript>
"""


def recaptcha_extractor(widget, data):
    request = data.request
    challenge_field = request.get('recaptcha_challenge_field')
    response_field = request.get('recaptcha_response_field')
    private_key = attr_value('private_key', widget, data)
    environ = getattr(request.request, 'environ', None)
    if not environ:
        environ = request.request
    remote_addr = environ.get('HTTP_X_FORWARDED_FOR', '').split(',')[0]
    if not remote_addr:
        remote_addr = environ.get('REMOTE_ADDR')
    res = submit(challenge_field, response_field, private_key, remote_addr)
    if not res.is_valid:
        raise ExtractionError(error_messages[res.error_code])
    return True


@managedprops('public_key', 'private_key', 'theme', 'lang')
def recaptcha_edit_renderer(widget, data):
    recaptcha_attrs = {
        'id': cssid(widget, 'recaptcha'),
        'class': ' '.join([cssclasses(widget, data)]),
    }
    data_attrs = ['theme', 'lang', 'public_key']
    recaptcha_attrs.update(data_attrs_helper(widget, data, data_attrs))
    recaptcha = data.tag('div', ' ', **recaptcha_attrs)
    public_key = attr_value('public_key', widget, data)
    return recaptcha + NO_SCRIPT_TEMPLATE.format(public_key=public_key)


def recaptcha_display_renderer(widget, data):
    pass


factory.register(
    'recaptcha',
    extractors=[recaptcha_extractor, generic_extractor],
    edit_renderers=[recaptcha_edit_renderer],
    display_renderers=[recaptcha_display_renderer])

factory.doc['blueprint']['recaptcha'] = """\
Add-on blueprint
`yafowil.widget.recaptcha <http://github.com/bluedynamics/yafowil.widget.recaptcha/>`_
"""

factory.defaults['recaptcha.class'] = 'recaptcha'

factory.defaults['recaptcha.error_class'] = 'error'

factory.defaults['recaptcha.message_class'] = 'errormessage'

factory.defaults['recaptcha.public_key'] = ''
factory.doc['props']['recaptcha.public_key'] = """\
reCAPTCHA public key. Used in the JavaScript code that is served to your users.
"""

factory.defaults['recaptcha.private_key'] = ''
factory.doc['props']['recaptcha.private_key'] = """\
reCAPTCHA private key. Used when communicating between your server and our
server. Be sure to keep it a secret.
"""

factory.defaults['recaptcha.lang'] = 'en'
factory.doc['props']['recaptcha.lang'] = """\
Language code.
"""

factory.defaults['recaptcha.theme'] = 'red'
factory.doc['props']['recaptcha.theme'] = """\
Used Theme. Available default themes are 'red' (Default), 'white', 'blackglass'
and 'clean'.
"""
